codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -54,16 +57,41 @@ import software.amazon.smithy.rust.codegen.core.util.inputShape
 import software.amazon.smithy.rust.codegen.core.util.outputShape
 import software.amazon.smithy.utils.StringUtils
 
+/**
+ * Class describing a JSON parser section that can be used in a customization.
+ */
+sealed class JsonParserSection(name: String) : Section(name) {
+    data class BeforeBoxingDeserializedMember(val shape: MemberShape) : JsonParserSection("BeforeBoxingDeserializedMember")
+}
+
+/**
+ * Customization for the JSON parser.
+ */
+typealias JsonParserCustomization = NamedSectionGenerator<JsonParserSection>
+
 class JsonParserGenerator(
-    private val codegenContext: CodegenContext,
+    codegenContext: CodegenContext,
     private val httpBindingResolver: HttpBindingResolver,
     /** Function that maps a MemberShape into a JSON field name */
     private val jsonName: (MemberShape) -> String,
+    /** Function that maps a StructureShape into its builder symbol */
+    private val builderSymbol: (StructureShape) -> Symbol,
+    /**
+     * Whether we should parse a value for a shape into its associated unconstrained type. For example, when the shape
+     * is a `StructureShape`, we should construct and return a builder instead of building into the final `struct` the
+     * user gets. This is only relevant for the server, that parses the incoming request and only after enforces
+     * constraint traits.
+     *
+     * The function returns a pair where the second component is the return symbol that should be parsed, and the first
+     * component is whether such symbol is unconstrained or not.
+     */
+    private val returnSymbolToParse: (Shape) -> Pair<Boolean, Symbol> = { shape -> false to codegenContext.symbolProvider.toSymbol(shape) },
```

Comment:
I think we should create a light `data class` to represent the pair so that the boolean and symbol can both be named.